File: apps/guzo-api/src/main/java/et/guzo/service/BranchStaffService.java
package et.guzo.service;

import et.guzo.common.ApiException;
import et.guzo.domain.entity.Branch;
import et.guzo.domain.entity.BranchStaff;
import et.guzo.repository.BranchRepository;
import et.guzo.repository.BranchStaffRepository;
import et.guzo.domain.entity.Role;
import et.guzo.domain.entity.UserRole;
import et.guzo.domain.entity.UserRoleId;
import et.guzo.repository.RoleRepository;
import et.guzo.repository.UserRoleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class BranchStaffService {

    private final BranchStaffRepository branchStaffRepository;
    private final RoleRepository roleRepository;
    private final UserRoleRepository userRoleRepository;
    private final BranchRepository branchRepository;

    @Transactional
    public BranchStaff assign(String userId, String branchId) {
        Role role = roleRepository.findByName("BRANCH_STAFF")
            .orElseThrow(() -> ApiException.badRequest("BRANCH_STAFF role is not configured"));
        UserRoleId roleKey = new UserRoleId(userId, role.getId());
        if (!userRoleRepository.existsById(roleKey)) {
            UserRole ur = new UserRole();
            ur.setUserId(userId);
            ur.setRoleId(role.getId());
            ur.setAssignedAt(Instant.now());
            userRoleRepository.save(ur);
        }

        BranchStaff staff = new BranchStaff();
        staff.setUserId(userId);
        staff.setBranchId(branchId);
        staff.setAssignedAt(Instant.now());
        return branchStaffRepository.save(staff);
    }

    @Transactional(readOnly = true)
    public List<BranchStaff> listByBranch(String branchId) {
        return branchStaffRepository.findByBranchId(branchId);
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> myBranches(String userId) {
        return branchStaffRepository.findByUserId(userId).stream().map(staff -> {
            Branch branch = branchRepository.findById(staff.getBranchId()).orElse(null);
            return Map.<String, Object>of(
                "branchId", staff.getBranchId(),
                "assignedAt", staff.getAssignedAt(),
                "branch", branch == null ? null : Map.of(
                    "id", branch.getId(),
                    "code", branch.getCode(),
                    "name", branch.getName(),
                    "city", branch.getCity(),
                    "line1", branch.getLine1(),
                    "phone", branch.getPhone()
                )
            );
        }).toList();
    }
}

File: apps/server/src/modules/branches/branch-staff.service.ts
import { prisma } from '@delivery/database';
import { ApiError } from '../../utils/ApiError.js';
import { ROLES } from '../../constants/index.js';

export class BranchStaffService {
  async assign(userId: string, branchId: string) {
    const branch = await prisma.guzoBranch.findUnique({ where: { id: branchId } });
    if (!branch) throw ApiError.notFound('Branch not found');

    const role = await prisma.role.findFirst({ where: { name: ROLES.BRANCH_STAFF } });
    if (!role) throw ApiError.badRequest('BRANCH_STAFF role is not configured');

    const existing = await prisma.userRole.findUnique({
      where: { userId_roleId: { userId, roleId: role.id } },
    });
    if (!existing) {
      await prisma.userRole.create({ data: { userId, roleId: role.id } });
    }

    return prisma.guzoBranchStaff.upsert({
      where: { userId_branchId: { userId, branchId } },
      create: { userId, branchId },
      update: {},
    });
  }

  async listByBranch(branchId: string) {
    return prisma.guzoBranchStaff.findMany({ where: { branchId } });
  }

  async myBranches(userId: string) {
    const rows = await prisma.guzoBranchStaff.findMany({
      where: { userId },
      include: { branch: true },
    });
    return rows.map((staff) => ({
      branchId: staff.branchId,
      assignedAt: staff.assignedAt,
      branch: staff.branch
        ? {
            id: staff.branch.id,
            code: staff.branch.code,
            name: staff.branch.name,
            city: staff.branch.city,
            line1: staff.branch.line1,
            phone: staff.branch.phone,
          }
        : null,
    }));
  }
}

export const branchStaffService = new BranchStaffService();
